evals/solvers/providers/qwen/qwen2audio_local_gpu_solver.py

```diff
@@ -0,0 +1,206 @@
+import base64
+import dataclasses
+import io
+import logging
+import queue
+import threading
+import time
+import traceback
+import torch
+import torch.distributed
+import torch.multiprocessing as mp
+import librosa
+import transformers
+from urllib.request import urlopen
+from concurrent import futures
+from concurrent.futures import ProcessPoolExecutor
+from typing import Any, Callable, Dict, List, Optional, TypeVar
+from evals.solvers.solver import Solver, SolverResult
+from evals.solvers.utils import BatchedProcessPoolExecutor
+from evals.task_state import TaskState
+
+SAMPLE_RATE = 16000
+DEFAULT_MAX_BATCH_SIZE = 32
+
+class Qwen2AudioLocalGPUSolver(Solver):
+    """
+    A solver class for running the Qwen2-Audio model in parallel across multiple GPUs.
+    Uses BatchedProcessPoolExecutor for efficient batch processing.
+
+    Args:
+        model_name: str - The model name/path (default: "Qwen/Qwen2-Audio-7B-Instruct")
+        num_gpus: int - Number of GPUs to use (default: all available)
+        max_batch_size: int - Maximum batch size for inference
+        extra_options: Dict[str, Any] - Additional options for model generation
+        postprocessors: list[str] - List of postprocessors to apply
+        registry: Any - Registry object for the solver
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        num_gpus: int = torch.cuda.device_count(),
+        max_batch_size: int = DEFAULT_MAX_BATCH_SIZE,
+        extra_options: Optional[Dict[str, Any]] = None,
+        postprocessors: list[str] = [],
+        registry: Any = None,
+    ):
+        super().__init__(postprocessors=postprocessors, registry=registry)
+
+        self.model_name = model_name
+        if extra_options is None:
+            extra_options = {}
+
+        # Set up multiprocessing
+        mp.set_start_method("spawn", force=True)
+        rank_queue = mp.Queue()
+        rank_queue.put(0)  # Start with primary GPU
+
+        self.executor = BatchedProcessPoolExecutor(
+            max_workers=max(1, num_gpus),
+            max_batch_size=int(max_batch_size),
+            initializer=solver_initializer,
+            initargs=(rank_queue, num_gpus, model_name),
+            batch_worker_fn=solver_worker,
+        )
+
+    def copy(self):
+        return self
+
+    def _process_audio_content(self, content: list) -> tuple[torch.Tensor, str]:
+        """Process audio content from message parts."""
+        audios = []
+        text_parts = []
+        
+        for i, part in enumerate(content):
+            if part["type"] == "audio_url":
+                if isinstance(part["audio_url"], dict) and "url" in part["audio_url"]:
+                    audio_data = part["audio_url"]["url"].split(",")[1]
+                    audio_stream = io.BytesIO(base64.b64decode(audio_data))
+                else:
+                    audio_stream = io.BytesIO(urlopen(part["audio_url"]).read())
+                
```

Comment:
nit: Huh, I don't think the local_gpu_solver handles the `urlopen` case. If you could refactor it so both can benefit from it, it would be great!